sched/sched/sched_note.c

```diff
@@ -631,144 +924,337 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
+             sizeof g_note_filter.syscall_mask);
+      for (i = 0; i < newf->nr_syscalls; i++)
+        {
+          nr = newf->syscalls[i];
 
-      tail = note_next(tail, 1);
-      remaining--;
+          if (nr < SYS_nsyscalls)
+            {
+              BITMASK_SET(g_note_filter.syscall_mask, nr);
+            }
+        }
     }
 
-  g_note_info.ni_tail = tail;
+  /* Count the number of filtered syscalls to calculate the requied size to
+   * get the current setting.
+   */
 
-errout_with_csection:
-  leave_critical_section(flags);
-  return notelen;
+  nr = 0;
+  for (i = 0; i < SYS_nsyscalls; i++)
+    {
+      if (BITMASK_CHECK(g_note_filter.syscall_mask, i))
+        {
+          nr++;
+        }
+    }
+
+  if (oldf != NULL)
+    {
+      /* Return the current filter setting */
+
+      int *p = oldf->syscalls;
+      int s = oldf->size;
+      oldf->nr_syscalls = nr;
+
+      for (i = 0; i < SYS_nsyscalls; i++)
+        {
+          if (p == NULL || s <= sizeof(int))
+            {
+              break;
+            }
+
+          if (BITMASK_CHECK(g_note_filter.syscall_mask, i))
+            {
+              *p++ = i;
+              s -= sizeof(int);
+            }
+        }
+    }
+
+  leave_critical_section(irq_mask);
 }
 #endif
 
 /****************************************************************************
- * Name: sched_note_size
+ * Name: sched_note_trace_irqfilter
  *
  * Description:
- *   Return the size of the next note at the tail of the circular buffer.
+ *   Set and get IRQ trace filter setting
+ *   (Same as TRIOC_GETIRQFILTER / TRIOC_SETIRQFILTER ioctls)
  *
  * Input Parameters:
- *   None.
+ *   oldf - A writable pointer to struct note_trace_irqfilter_s to get
+ *          current IRQ trace filter setting
+ *          If 0, no data is written.
+ *   newf - A read-only pointer to struct note_trace_irqfilter_s of the new
+ *          IRQ trace filter setting
+ *          If 0, the setting is not updated.
  *
  * Returned Value:
- *   Zero is returned if the circular buffer is empty.  Otherwise, the size
- *   of the next note is returned.
+ *   None
  *
  ****************************************************************************/
 
-#ifdef CONFIG_SCHED_NOTE_GET
-ssize_t sched_note_size(void)
+#ifdef CONFIG_SCHED_INSTRUMENTATION_IRQHANDLER
+void sched_note_trace_irqfilter(struct note_trace_irqfilter_s *oldf,
+                                struct note_trace_irqfilter_s *newf)
 {
-  FAR struct note_common_s *note;
-  irqstate_t flags;
-  unsigned int tail;
-  ssize_t notelen;
-  size_t circlen;
+  irqstate_t irq_mask;
+  int i;
+  int nr;
 
-  flags = enter_critical_section();
+  irq_mask = enter_critical_section();
 
-  /* Verify that the circular buffer is not empty */
+  if (newf != NULL)
+    {
+      sched_note_trace_stop();
+
+      /* Replace the IRQ filter mask by the provided setting */
+
+      memset(g_note_filter.irqhandler_mask, 0,
+             sizeof g_note_filter.irqhandler_mask);
```

Comment:
should we use the same format here to avoid the conversion(number<->bitmap)?

Reply:
Same as syscallfiter.